pageserver/src/tenant.rs

```diff
@@ -1231,11 +1234,22 @@ impl Tenant {
             .unwrap_or_else(|| "-".to_string());
 
         {
-            let _timer = STORAGE_TIME
+            let timer = Instant::now();
+            let gc_result = self
+                .gc_iteration_internal(target_timeline_id, horizon, pitr)
+                .await;
+            let duration = timer.elapsed().as_secs_f64();
+            STORAGE_TIME_GLOBAL
+                .with_label_values(&["gc"])
+                .observe(duration);
+            STORAGE_TIME_SUM_PER_TIMELINE
                 .with_label_values(&["gc", &self.tenant_id.to_string(), &timeline_str])
-                .start_timer();
-            self.gc_iteration_internal(target_timeline_id, horizon, pitr)
-                .await
+                .inc_by(duration);
+            STORAGE_TIME_COUNT_PER_TIMELINE
+                .with_label_values(&["gc", &self.tenant_id.to_string(), &timeline_str])
```

Comment:
Can we get rid of these `with_label_values()` calls? If the GC loop outlives the `TimelineMetrics`, this will recreate the metrics even though the timeline is deleted.

Reply:
Good question, didn't even think about the possibility. Did you have something already in mind regarding that? Not sure yet how would I approach that.

Reply:
> If the GC loop outlives the `TimelineMetrics`

Unsure if this can happen. Do we have a test case for this? refresh_gc_info would have to pick up the timeline, which would then have to be detached before `timeline.gc().await` is called.. I can see the detach easily happening in between, but the same `Arc<Timeline>` would keep the TimelineMetrics alive, surely?

Reply:
> Do we have a test case for this? 

Not really, but for me,  `tenant_metrics.py` uncovered missing metric cleanup in the past.

Reply:
> Not really, but for me, tenant_metrics.py uncovered missing metric cleanup in the past.
Do you have any details on how could I reproduce that? I mean we didn't have any kind of cleanup in the past.

> I can see the detach easily happening in between, but the same Arc<Timeline> would keep the TimelineMetrics alive, surely?

That sounds like making sense to me.

Not sure if it would make any difference regarding the behavior, but the original point about calling `with_label_values()` - or perhaps rather initializing the metric here - perhaps seems a bit odd anyway, considering this is the only operation that doesn't go through `TimelineMetrics`. I think we could handle the metrics in `gc_iteration_internal`. That would effectively mean though, that we would drop the metric for garbage collecting all the timelines (`timeline_id="-"`). Unsure how useful is that. But at the same time it could start to be a bit of scope creep in context of the purpose of this PR.

What do you think?

Reply:
So would look like https://github.com/neondatabase/neon/pull/3409/commits/2f1f8c4ebb55458be31a212124ee6e02307cd013, just that would probably make sense to unnest the gc call in that case.

Reply:
Yeah, that's better.

Regarding the `{timeline_id="-"}` case:

if`cfg(not(testing))`, is `target_timeline_id` ever `Some()`? If it's always `None` in production, are the per_timeline times actually interesting to us?

Or, conversely, we could have a separate histogram that just groups by `"gc", tenant_id`

Reply:
Thank you for pointing this out! I had completely missed the fact that there are no timeline specific ones in production. Probably because all the rest are per timeline (wonder how important that is). Keeping that in mind, I wonder if it would then be a bit confusing to make the call via timeline metrics.

Reply:
>  are the per_timeline times actually interesting to us?

Another thing I previously didn't really think about is that we seem to garbage collect the timelines sequentially: https://github.com/neondatabase/neon/blob/bf63f129ae1b1822dfb2ec689e483f39c772c9a5/pageserver/src/tenant.rs#L1921-L1929. The timer is probably dependent on how many timelines there are, so it's actually a bit unclear to me what are we after with this metric.

That being said, I would think in production use case we actually get a bit more information with the new "summary", as we can `increment(gc_time) / increment(gc_time_count)`. So alleviating the possibly changing timeline count.

So back to the question, I didn't find any stored metrics where there would be `gc` statistics per timeline_id, so that at least supports just getting rid of it.

Reply:
The current state makes more sense to me.

Reply:
Moved this to timeline not to have tenant touch timelines privates. Resolving.